### PR TITLE
test: address edge cases for e2e secret rotation

### PR DESCRIPTION
**Pull request by sdowell**
This change adds the GCP iam needed by the e2e test runner GSA, and adds
some logic to harden the secret rotation for certain edge cases.

Builds on https://github.com/GoogleContainerTools/kpt-config-sync/pull/1077

### CODE REVIEW

File: e2e/testinfra/terraform/prow/service_accounts.tf [hunk 1]
```diff
@@ -23,11 +23,17 @@ resource "google_project_iam_member" "test-runner-iam" {
...
 
   role    = each.value
   member  = "serviceAccount:e2e-test-runner@oss-prow-build-kpt-config-sync.iam.gserviceaccount.com"
   project = data.google_project.project.id
 }
+
+resource "google_service_account_iam_member" "admin-account-iam" {
+  service_account_id = "projects/${data.google_project.project.id}/serviceAccounts/e2e-test-ar-reader@${data.google_project.project.name}.iam.gserviceaccount.com"
+  role               = "roles/iam.serviceAccountKeyAdmin"
+  member             = "serviceAccount:e2e-test-runner@oss-prow-build-kpt-config-sync.iam.gserviceaccount.com"
```
Comment (karlkfi) [inline e2e/testinfra/terraform/prow/service_accounts.tf:38]: should the project id be interpolated from the project var?

Comment (sdowell) [reply e2e/testinfra/terraform/prow/service_accounts.tf:38]: Not for this service account, because the prow project and the test-runner SA are singletons

**Comment by karlkfi** [comment]
You can rebase on https://github.com/GoogleContainerTools/kpt-config-sync/pull/1077 now

File: e2e/testcases/helm_sync_test.go [hunk 2]
```diff
@@ -935,14 +942,27 @@ func fetchServiceAccountKeyFile(nt *nomostest.NT, projectID, gsaKeySecretID, gsa
...
+			return "", err
+		}
+		return gsaKeyFilePath, nil
 	}
 	// For some reason, the "name" value is just the version,
 	// even tho in json the "name" field is the full name.
 	// But what the access command wants is the version. So it's fine.
 	gsaKeySecretVersion := strings.TrimSpace(string(out))
-	if gsaKeySecretVersion == "" {
+	if gsaKeySecretVersion == "" || strings.Contains(gsaKeySecretVersion, "WARNING") {
```
Comment (karlkfi) [inline e2e/testcases/helm_sync_test.go:965]: what's the warning edge case?

Comment (sdowell) [reply e2e/testcases/helm_sync_test.go:965]: If you list a secret that exists but doesn't have any versions, it prints a warning message but doesn't return an error

Comment (sdowell) [reply e2e/testcases/helm_sync_test.go:965]: Another approach could be trying to parse the output to an integer and failing if it errors

Comment (karlkfi) [reply e2e/testcases/helm_sync_test.go:965]: Hmm how about matching on lines that start with WARNING and removing the whole line, just in case there's some other warning we don't know about. 

Or check for exact match of the empty warning and error if there are other unexpected warnings.

Comment (sdowell) [reply e2e/testcases/helm_sync_test.go]: I don't want this to be too brittle, as we're just checking a warning string on the output for errors. If the warning message changes slightly, I'd prefer this not break. 

I updated this to pass in the `--no-user-output-enabled` flag, which suppresses the warning message. This means that when there are no versions the output should more reliably pass the empty check on this line.

Comment (karlkfi) [reply e2e/testcases/helm_sync_test.go]: great, thanks.

File: e2e/testinfra/terraform/prow/service_accounts.tf [hunk 3]
```diff
@@ -23,11 +23,17 @@ resource "google_project_iam_member" "test-runner-iam" {
...
     "roles/source.reader",
   ])
 
   role    = each.value
   member  = "serviceAccount:e2e-test-runner@oss-prow-build-kpt-config-sync.iam.gserviceaccount.com"
   project = data.google_project.project.id
 }
+
+resource "google_service_account_iam_member" "admin-account-iam" {
+  service_account_id = "projects/${data.google_project.project.id}/serviceAccounts/e2e-test-ar-reader@${data.google_project.project.name}.iam.gserviceaccount.com"
```
Comment (karlkfi) [inline e2e/testinfra/terraform/prow/service_accounts.tf:36]: Why name and not id? Isn't it the ID in the GSA email?

File: e2e/testinfra/terraform/prow/service_accounts.tf [hunk 4]
```diff
@@ -23,11 +23,17 @@ resource "google_project_iam_member" "test-runner-iam" {
     "roles/container.admin",
     "roles/gkehub.admin",
     "roles/monitoring.viewer",
-    "roles/secretmanager.secretAccessor",
+    "roles/secretmanager.admin",
```
Comment (nan-yu) [inline e2e/testinfra/terraform/prow/service_accounts.tf:26]: Do we need the `secretmanager.admin` role? The only missing role is `secretmanager.viewer`. We can have both `secretmanager.secretAccessor` and `secretmanager.viewer`.

**Review by nan-yu — COMMENTED** [review]
Thanks for fixing the TF config for the missing role.

File: e2e/testinfra/terraform/prow/service_accounts.tf (hunk 4 above)
Comment (karlkfi) [reply e2e/testinfra/terraform/prow/service_accounts.tf:26]: The code destroys secrets when they're invalid and creates them when they don't exist. So it probably needs admin.

**Review by karlkfi — APPROVED** [review]
/lgtm

File: e2e/testinfra/terraform/prow/service_accounts.tf (hunk 4 above)
Comment (sdowell) [reply e2e/testinfra/terraform/prow/service_accounts.tf:26]: With https://github.com/GoogleContainerTools/kpt-config-sync/pull/1077 the SA needs a broader range of permissions. It needs permission now to add/disable secret versions when the key expires, and potentially create a new secret if running on a new environment that needs bootstrapping. We could remove the ability to bootstrap the initial secret if we want to keep the permissions more minimal, although that means it needs to be done manually when creating a new test environment.

Comment (nan-yu) [reply e2e/testinfra/terraform/prow/service_accounts.tf:26]: ACK

**Review by nan-yu — APPROVED** [review]
/lgtm
